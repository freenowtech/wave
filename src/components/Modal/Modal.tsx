import React, { useEffect, useState, ReactNode, useContext } from 'react';
import { createGlobalStyle } from 'styled-components';
import { WidthProps } from 'styled-system';
import { useIsEscKeyPressed } from '../../utils/hooks/useIsEscKeyPressed';
import { ANIMATION_DURATION as CARD_ANIMATION_DURATION, CenteredCard } from './components/CenteredCard';
import { ANIMATION_DURATION as DIMMING_ANIMATION_DURATION, DimmingFade } from './components/DimmingFade';
import { TopRightXIcon } from './components/TopRightXIcon';

type DismissFunc = () => void;

const DismissContext = React.createContext<DismissFunc>(undefined);

const useModalDismiss = (): DismissFunc => {
    const dismiss = useContext(DismissContext);

    if (dismiss === undefined) {
        throw new Error('useModalDismiss must be used within a <Modal />');
    }

    return dismiss;
};

interface ModalProps extends WidthProps {
    /**
     * Show the modal covering the whole page to focus the users attention
     */
    fullscreen?: boolean;
    /**
     * Makes the modal dismissible by the user (defaults to true)
     */
    dismissible?: boolean;
    /**
     * Function that will be called when the user intends to close the modal
     */
    onClose?: () => void;
    /**
     * Children can be rendered with a dismiss function as render prop. Calling
     * the dismiss function will close the Modal.
     */
    children?: ReactNode | ((dismiss: DismissFunc) => ReactNode);
}

const PreventBackgroundScroll = createGlobalStyle`
    body {
        overflow: hidden;
    }
`;

const ANIMATION_DURATION = Math.max(DIMMING_ANIMATION_DURATION, CARD_ANIMATION_DURATION);

/*
 * We define the types for the props twice because the render props won't be inferred correctly
 * when only using `React.FC<ModalProps>`. This leads to compiler errors when passing the
 * dismiss function.
 */
const Modal: React.FC<ModalProps> = ({ children, onClose, dismissible, ...rest }: ModalProps) => {
    const [visible, setVisible] = useState(true);
    const isEscKeyPressed = useIsEscKeyPressed();

    const handleClose: DismissFunc = () => {
        setVisible(false);

        if (onClose) {
            setTimeout(() => onClose(), ANIMATION_DURATION);
        }
    };

    const handleDimmingClick = () => {
        if (dismissible && !rest.fullscreen) {
            handleClose();
        }
    };

    useEffect(() => {
        if (dismissible && isEscKeyPressed) {
            handleClose();
        }
    }, [dismissible, isEscKeyPressed]);

    const renderChildren = () => {
        if (typeof children === 'function') {
            return children(handleClose);
        }

        return children;
    };

    return (
        <DismissContext.Provider value={handleClose}>
            <DimmingFade
                onClick={handleDimmingClick}
                visible={visible}
                fullscreen={rest.fullscreen}
                data-testid="dimming-background"
            >
                {dismissible && rest.fullscreen && <TopRightXIcon data-testid="close-icon" onClick={handleClose} />}
            </DimmingFade>
            <CenteredCard visible={visible} {...rest}>
                {dismissible && !rest.fullscreen && <TopRightXIcon data-testid="close-icon" onClick={handleClose} />}
                {renderChildren()}
            </CenteredCard>
            <PreventBackgroundScroll />
        </DismissContext.Provider>
    );
};

Modal.defaultProps = {
    dismissible: true
};

export { Modal, ModalProps, useModalDismiss };
