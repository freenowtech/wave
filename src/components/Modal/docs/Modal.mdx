---
name: Modal
menu: Components
route: /components/Modal
---

import { Modal } from '../Modal';
import { ModalCreator } from './ModalCreator';
import { ModalPropsTable } from './ModalPropsTable';
import { StyledSystemLinks } from '../../../docs/StyledSystemLinks'

# Modal

The Modal should only contain content or flows which are related to the current page and not replace pages, therefore
its complexity level has to be simple. Consider the fullscreen state if you want to focus the users attention to the modal
content.

Modal is responsive per default but the width can be adjusted (the default is 600px or 37.5rem). Content on the modal
is scrollable and has a space5 bounding box.

<ModalCreator />

### Example Code

The following contains a simple example, with a resulting modal similar to the live examples above.

```jsx
const [showModal, setShowModal] = useState(true);

<Modal onClose={() => setShowModal(false)} fullscreen={true}>
    {dismiss => (
        <>
            <Headline as="h2">Add Note</Headline>
            <Text as="p" my={3}>
                Ensure that all participants are aware of the change.
            </Text>
            <Button onClick={dismiss}>Add Note</Button>
            <TextButton onClick={dismiss}>Cancel</TextButton>
        </>
    )}
</Modal>;
```

## API

<ModalPropsTable />

<br />

<StyledSystemLinks component="Modal" supportedProps={ ['width'] }/>

### Dismiss

Modal provides a function to dismiss the component with an animation. The
dismiss function is available either through render props or a hook. After the
animation has finished the `onClose` callback will be called.

#### Render Props

```jsx
<Modal>
    {dismiss => <Button onClick={dismiss}>Close</Button>}
</Moal>
```

#### Hook

```jsx
const InsideModal = () => {
    const dismiss = useModalDismiss();
    return <button onClick={dismiss}>Close</button>;
}

const ModalWrapper = () => (
    <Modal>
        <InsideModal />
    </Modal>
);
```
