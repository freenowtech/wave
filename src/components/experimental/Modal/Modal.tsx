import React from 'react';
import styled from 'styled-components';
import { Dialog, DialogProps, Modal as BaseModal } from 'react-aria-components';
import { getSemanticValue } from '../../../essentials/experimental';

const ModalCard = styled(BaseModal)`
    padding: 2rem;
    border-radius: 1.5rem;
    background: ${getSemanticValue('surface')};
    color: ${getSemanticValue('on-surface')};
    outline: none;

    &[data-entering] {
        animation: modal-zoom 300ms cubic-bezier(0.175, 0.885, 0.32, 1.275);
    }

    @keyframes modal-zoom {
        from {
            transform: scale(0.8);
        }
        to {
            transform: scale(1);
        }
    }
`;

const StyledDialog = styled(Dialog)`
    outline: none;
`;

type ModalProps = Pick<DialogProps, 'children' | 'role'>;

const Modal = React.forwardRef<HTMLDivElement, ModalProps>((props, ref) => (
    <ModalCard ref={ref}>
        <StyledDialog {...props} />
    </ModalCard>
));

export { Modal, ModalProps };
