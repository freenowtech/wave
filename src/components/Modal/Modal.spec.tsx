import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as React from 'react';
import { Modal, useModalDismiss } from './Modal';

describe('Modal', () => {
    it('renders the default props', () => {
        expect(render(<Modal>Content</Modal>).container).toMatchSnapshot();
    });

    it('renders the fullscreen variation', () => {
        expect(render(<Modal fullscreen>Content</Modal>).container).toMatchSnapshot();
    });

    describe('should call onClose function', () => {
        it('when clicking on close icon', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(<Modal onClose={mockCloseHandler}>Content</Modal>);

            fireEvent.click(getByTestId('close-icon'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when clicking on close icon in fullscreen', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(
                <Modal onClose={mockCloseHandler} fullscreen>
                    Content
                </Modal>
            );

            fireEvent.click(getByTestId('close-icon'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when clicking on the dimming background', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(<Modal onClose={mockCloseHandler}>Content</Modal>);

            fireEvent.click(getByTestId('dimming-background'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when pressing the escape key', async () => {
            const mockCloseHandler = jest.fn();
            render(<Modal onClose={mockCloseHandler}>Content</Modal>);

            fireEvent.keyUp(window, { keyCode: 27 });

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when calling the dismiss function', async () => {
            const mockCloseHandler = jest.fn();
            const { getByText } = render(
                <Modal onClose={mockCloseHandler}>
                    {dismiss => (
                        <button type="button" onClick={dismiss}>
                            dismiss
                        </button>
                    )}
                </Modal>
            );

            fireEvent.click(getByText('dismiss'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when calling the dismiss function from a hook', async () => {
            const mockCloseHandler = jest.fn();

            const InnerComponent: React.FC = () => {
                const dismiss = useModalDismiss();
                return (
                    <button type="button" onClick={dismiss}>
                        Click Me
                    </button>
                );
            };

            render(
                <Modal onClose={mockCloseHandler}>
                    <InnerComponent />
                </Modal>
            );

            userEvent.click(screen.getByRole('button'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });
    });

    describe('in the non-dismissible state', () => {
        it('should not show the close icon for the default modal', () => {
            const { queryByTestId } = render(<Modal dismissible={false} />);

            expect(queryByTestId('close-icon')).toBeNull();
        });

        it('should not show the close icon for the fullscreen modal', () => {
            const { queryByTestId } = render(<Modal fullscreen dismissible={false} />);

            expect(queryByTestId('close-icon')).toBeNull();
        });

        it('should not call the onClose handler when clicking on the dimming', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(<Modal dismissible={false} onClose={mockCloseHandler} />);

            fireEvent.click(getByTestId('dimming-background'));

            await waitFor(() => expect(mockCloseHandler).not.toHaveBeenCalled());
        });

        it('should not call the onClose handler when pressing the escape key', async () => {
            const mockCloseHandler = jest.fn();
            render(
                <Modal dismissible={false} onClose={mockCloseHandler}>
                    Content
                </Modal>
            );

            fireEvent.keyUp(window, { keyCode: 27 });

            await waitFor(() => expect(mockCloseHandler).not.toHaveBeenCalled());
        });

        it('should call the onClose handler when calling the dismiss function', async () => {
            const mockCloseHandler = jest.fn();
            const { getByText } = render(
                <Modal onClose={mockCloseHandler} dismissible={false}>
                    {dismiss => (
                        <button type="button" onClick={dismiss}>
                            dismiss
                        </button>
                    )}
                </Modal>
            );

            fireEvent.click(getByText('dismiss'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });
    });

    it('should not call onClose function when clicking dimming in fullscreen', async () => {
        const mockCloseHandler = jest.fn();
        const { getByTestId } = render(
            <Modal onClose={mockCloseHandler} fullscreen>
                Content
            </Modal>
        );

        fireEvent.click(getByTestId('dimming-background'));

        await waitFor(() => expect(mockCloseHandler).not.toHaveBeenCalled());
    });
});
