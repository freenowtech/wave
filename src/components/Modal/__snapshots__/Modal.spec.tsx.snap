// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders the default props 1`] = `
.c2 {
  overflow: auto;
  box-sizing: border-box;
  background-color: #FFFFFF;
  border-radius: 0.125rem;
  padding: 2.5rem;
  width: 37.5rem;
  box-shadow: 0 0 0.75rem 0.25rem rgba(0,0,0,0.12);
}

.c3 {
  position: fixed;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1100;
  max-height: calc(100% - 2rem);
  will-change: transform,opacity;
  -webkit-transition: -webkit-transform 150ms ease-out,opacity 150ms ease-out;
  -webkit-transition: transform 150ms ease-out,opacity 150ms ease-out;
  transition: transform 150ms ease-out,opacity 150ms ease-out;
}

.c3.centered-card-animation-appear,
.c3.centered-card-animation-enter {
  opacity: 0;
  -webkit-transform: translate(-50%,calc(-50% - 1rem));
  -ms-transform: translate(-50%,calc(-50% - 1rem));
  transform: translate(-50%,calc(-50% - 1rem));
}

.c3.centered-card-animation-appear-active,
.c3.centered-card-animation-enter-active {
  opacity: 1;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c3.centered-card-animation-exit {
  opacity: 1;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c3.centered-card-animation-exit-done,
.c3.centered-card-animation-exit-active {
  opacity: 0;
  -webkit-transform: translate(-50%,-50%) scale(0.98);
  -ms-transform: translate(-50%,-50%) scale(0.98);
  transform: translate(-50%,-50%) scale(0.98);
}

.c0 {
  background-color: rgba(0,15,31,0.6);
  height: 100%;
  left: 0;
  opacity: 0.6;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1000;
}

.c1 {
  will-change: opacity;
  -webkit-transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
  transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
}

.c1.dimming-fade-animation-appear,
.c1.dimming-fade-animation-enter {
  opacity: 0;
}

.c1.dimming-fade-animation-appear-active,
.c1.dimming-fade-animation-enter-active {
  opacity: 0.6;
}

.c1.dimming-fade-animation-exit {
  opacity: 0.6;
}

.c1.dimming-fade-animation-exit-done,
.c1.dimming-fade-animation-exit-active {
  opacity: 0;
}

.c4 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  z-index: 1100;
}

@media (max-width:calc(37.5rem + 2rem)) {
  .c3 {
    width: calc(100% - 2rem);
  }
}

<div>
  <div
    class="c0 c1 dimming-fade-animation-appear dimming-fade-animation-appear-active"
    data-testid="dimming-background"
  />
  <div
    class="c2 c3 centered-card-animation-appear centered-card-animation-appear-active"
    width="37.5rem"
  >
    <svg
      class="c4"
      color="#001E3E"
      data-testid="close-icon"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        fill="currentColor"
        fill-rule="nonzero"
      />
    </svg>
    Content
  </div>
</div>
`;

exports[`Modal renders the fullscreen variation 1`] = `
.c3 {
  overflow: auto;
  box-sizing: border-box;
  background-color: #FFFFFF;
  border-radius: 0.125rem;
  padding: 2.5rem;
  width: 37.5rem;
  box-shadow: none;
}

.c4 {
  position: fixed;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1100;
  max-height: calc(100% - 2rem);
  will-change: transform,opacity;
  -webkit-transition: -webkit-transform 150ms ease-out,opacity 150ms ease-out;
  -webkit-transition: transform 150ms ease-out,opacity 150ms ease-out;
  transition: transform 150ms ease-out,opacity 150ms ease-out;
}

.c4.centered-card-animation-appear,
.c4.centered-card-animation-enter {
  opacity: 0;
  -webkit-transform: translate(-50%,-50%) scale(0.94);
  -ms-transform: translate(-50%,-50%) scale(0.94);
  transform: translate(-50%,-50%) scale(0.94);
}

.c4.centered-card-animation-appear-active,
.c4.centered-card-animation-enter-active {
  opacity: 1;
  -webkit-transform: translate(-50%,-50%) scale(1);
  -ms-transform: translate(-50%,-50%) scale(1);
  transform: translate(-50%,-50%) scale(1);
}

.c4.centered-card-animation-exit {
  opacity: 1;
  -webkit-transform: translate(-50%,-50%) scale(1);
  -ms-transform: translate(-50%,-50%) scale(1);
  transform: translate(-50%,-50%) scale(1);
}

.c4.centered-card-animation-exit-done,
.c4.centered-card-animation-exit-active {
  opacity: 0;
  -webkit-transform: translate(-50%,-50%) scale(0.98);
  -ms-transform: translate(-50%,-50%) scale(0.98);
  transform: translate(-50%,-50%) scale(0.98);
}

.c0 {
  background-color: rgba(0,15,31,0.6);
  height: 100%;
  left: 0;
  opacity: 0.6;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1000;
}

.c1 {
  will-change: opacity;
  -webkit-transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
  transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
  background-color: #FFFFFF;
  opacity: 1;
}

.c1.dimming-fade-animation-appear,
.c1.dimming-fade-animation-enter {
  opacity: 0;
}

.c1.dimming-fade-animation-appear-active,
.c1.dimming-fade-animation-enter-active {
  opacity: 1;
}

.c1.dimming-fade-animation-exit {
  opacity: 1;
}

.c1.dimming-fade-animation-exit-done,
.c1.dimming-fade-animation-exit-active {
  opacity: 0;
}

.c2 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  z-index: 1100;
}

@media (max-width:calc(37.5rem + 2rem)) {
  .c4 {
    width: calc(100% - 2rem);
  }
}

<div>
  <div
    class="c0 c1 dimming-fade-animation-appear dimming-fade-animation-appear-active"
    data-testid="dimming-background"
  >
    <svg
      class="c2"
      color="#001E3E"
      data-testid="close-icon"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        fill="currentColor"
        fill-rule="nonzero"
      />
    </svg>
  </div>
  <div
    class="c3 c4 centered-card-animation-appear centered-card-animation-appear-active"
    width="37.5rem"
  >
    Content
  </div>
</div>
`;
